Feature(rest)(thrift): introduction of healthcheck

neubs-bsi opened the pull request:
[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

> Please provide a summary of your changes here.
> * Which issue is this pull request belonging to and how is it solving it? (*Refer to issue here*)
> * Did you add or update any new dependencies that are required for your change?

This PR introduces a basic health check to SW360 rest-api and backend.

With the introduction of the spring boot actuator a health endpoint was created.
With a health indicator the health check specifically checks the db connection to the attachment database and the connection to a new thrift client that is specific for health. 

That thrift client has a getHealth() method that checks the connection of the thrift client to the other databases.
Should that check throw a TException, it is known, that the thrift client is not reachable. 

Issue: https://github.com/eclipse/sw360/issues/896

New dependencies:
- org.springframework.boot:spring-boot-starter-actuator:${spring-boot.version}

### Suggest Reviewer
> You can suggest reviewers here with an @mention.

### How To Test?
> How should these changes be tested by the reviewer?
> Have you implemented any additional tests?
Integration and Unit tests were added. 

### Checklist
Must:
- [x] All related issues are referenced in commit messages and in PR

mcjaeger commented:
Just a question since we have had this in our telcos: is it that each commit makes separately? Again, we would to be able to conveniently cherry pick commits for easy handling of forks ...

neubs-bsi commented:
I believe all commits have something they introduce or change that validates an own commit, but if it is of concern, I can squash them into more broad themed commits.

neubs-bsi commented:
From another perspective, it is also possible to use the "squash and merge" button when merging. This way the review could be done with my 17 commits, which makes it easier to understand what I have done and when and then have a squashed commit in total for the final cherry picking.

mcjaeger commented:
@neubs-bsi absolutely was having the same idea later too (to squash commits at merge PR time). so no problem,...

neubs-bsi commented on rest/resource-server/pom.xml:
```diff
@@ -28,6 +29,12 @@
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-actuator</artifactId>
+            <scope>compile</scope>
```
I was thinking the same, but got thrown of by the rest of the pom file and thought I'd try to adjust to the style of the document. I will adjust it. 
Is there a reason that some other components in the same pom file have the `compile` scope?

neubs-bsi commented on rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/SW360RestHealthIndicator.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    private boolean isThriftReachable(List<Exception> exception) {
+        HealthService.Iface healthClient = makeHealthClient();
```
I am doing it the way it is done in the rest of the code, except that I extracted the call `new ThriftClients().makeHealthClient()` into a separate function.

Would you like me to not do it in a separate method?

JaideepPalit replied:
The code is fine.
I was trying to say, since thrift client wont change, maybe we can create a single `HealthService.Iface` object and reuse it.
Instead of creating a new one with every request.

```
    private static HealthService.Iface healthService;
    protected HealthService.Iface makeHealthClient() {
        if(healthService==null) {
            healthService = new ThriftClients().makeHealthClient();
        }
        return healthService;
    }
```

neubs-bsi replied:
I was under the impression that this is not possible due to Thrift not being threadsafe.
Since the calls are made via Spring we can't control the threads. Hence the client could be corrupted.
Since I am not sure what exactly thrift does under the hood to its client in a thread I would opt for the safest option, which is to declare a new one for each thread.

neubs-bsi commented on rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/SW360RestHealthIndicator.java:
```diff
@@ -0,0 +1,107 @@
+                return true;
+            } else {
+                exception.add(
+                        new Exception(health.getStatus().toString(),
```
The Exception here is a formality that is needed because in this version of the spring boot actuator it does not take throwables in its exception status. The actual exception will unfortunately be wrapped within this exception already. 

I don't think making this a custom exception adds information to the exception, which should be the goal of custom exceptions.

neubs-bsi commented:
Thanks for the review. I have rebased and included most of your review comments. 
To the log statements in the start I am still checking if I can recreate them, so I will get back to you on that.

JaideepPalit commented:
> It would be good to have documentation for the health APIs.
    - Any thoughts regarding this comment?

Re-Deployed the PR in Java 11 - build and deploy is successfull. Resource-server is launching and able to make simple rest calls like GetProjectById.

- Tested `/health`
   - with DB running and shutdown - working fine
   - deleted health.war from tomcat - status shows down - working fine
   - deleting only datahandler from LR-OSGI- status still show UP. However loading of project stops
   - response doesn't contain details, however added in code

_**SW360RestHealthIndicator.java - Line 44**_
![image](https://user-images.githubusercontent.com/56534710/91854230-0b22fc80-ec81-11ea-8359-730795b5b8f5.png)

![image](https://user-images.githubusercontent.com/56534710/91854154-edee2e00-ec80-11ea-958d-77b148146065.png)

- While accessing `/info` getting 401 even after adding valid token

Could you provide some Steps on **How to test**  and **expected behavior** of this PR?

neubs-bsi commented:
@JaideepPalit thanks again for the thorough review. 

Since the main health functionality comes from the spring boot actuator, the configuration of the endpoints is also done there. 

> While accessing /info getting 401 even after adding valid token

Since the accessibility of endpoints giving information about the application can be a security risk, I disabled everything other that the simple health information by default. This is detailed [here](https://docs.spring.io/spring-boot/docs/1.5.3.RELEASE/reference/htmlsingle/#production-ready-sensitive-endpoints). 

I have added a documentation for the endpoint. With this snippet in the default application yaml:

```
endpoints:
  enabled: false
  info:
    enabled: true
    sensitive: true
  health:
    enabled: true
    sensitive: false

```

You can see, that the endpoint for `info` was set to `sensitiv`. This would require authentication with the role `ACTUATOR` to access the `info` endpoint. This does not exist within SW360 and would need to be configured specifically to work. I "disabled" the `info` endpoint, because I reconned this PR is about the `health` endpoint and introducing the `info` endpoint by default seemed invasive. In this version of the actuator the setting `sensitiv` for the endpoints `health` and `info` are `false` by default. 

> deleting only datahandler from LR-OSGI- status still show UP. However loading of project stops

Since this healthcheck is specifig for the rest case, the Liferay connection or health is not checked. This would be an additional feature. So when you delete the datahandler, the health status will not change. 

> response doesn't contain details, however added in code

The details given also need to be configured in the `application.yml`. In the `application.properties` of the test resources, you can see that I added `management.security.enabled=false`. This enables a detailed description of the health status. Since this setting also has security implication for other spring features I did not want to put it into the default. 

So here is the **expected behaviour**:

With this setup: 
```
endpoints:
  enabled: false
  info:
    enabled: true
    sensitive: true
  health:
    enabled: true
    sensitive: false

```
- When CouchDB is shutdown -> status is `DOWN`
- When any of the databases (sw360db, sw360attachements, sw360users) in the couchdb are missing -> status is `DOWN`
- When Thrift Server `health` is shut down / deleted -> status is `DOWN`
- When the spring server is down -> status is `DOWN`
- endpoint `info` is not reachable
- only the status information is provided in the health call
- changes to the frontend do not have an influence to the health status
- missing databases, other than the above mentioned, do not have an influence to the health status

With the more permissive setup:

```
endpoints:
  enabled: false
  info:
    enabled: true
    sensitive: false
  health:
    enabled: true
    sensitive: false
manager:
  security:
    enabled: false

```
- endpoint `info` is reachable
- in the health call the status and details are provided in the respone

JaideepPalit reviewed (approved):
Code looks fine to me. Tested successfully.

JaideepPalit commented:
Waiting of travis build to pass , to merge.
